pkg/client/combined/combined.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2019 The GitLab-Controller Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package combined
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Client is a client.Client that updates the metadata, spec, and status of
+// objects using the status subresource with one Update call. All other calls
+// are passed directly to the underlying client.Client. Client should only be
+// used with resources that have +kubebuilder:subresource:status enabled.
+type Client struct {
+	c client.Client
+}
+
+// NewClient wraps the supplied Client with a Client client.
+func NewClient(c client.Client) *Client {
+	return &Client{c: c}
+}
+
+// Get retrieves an obj for the given object key from the Kubernetes Cluster.
+// obj must be a struct pointer so that obj can be updated with the response
+// returned by the Server.
+func (c *Client) Get(ctx context.Context, key client.ObjectKey, obj runtime.Object) error {
+	return c.c.Get(ctx, key, obj)
+}
+
+// List retrieves list of objects for a given namespace and list options. On a
+// successful call, Items field in the list will be populated with the
+// result returned from the server.
+func (c *Client) List(ctx context.Context, opts *client.ListOptions, list runtime.Object) error {
+	return c.c.List(ctx, opts, list)
+}
+
+// Create saves the object obj in the Kubernetes cluster.
+func (c *Client) Create(ctx context.Context, obj runtime.Object) error {
+	return c.c.Create(ctx, obj)
+}
+
+// Delete deletes the given obj from Kubernetes cluster.
+func (c *Client) Delete(ctx context.Context, obj runtime.Object, opts ...client.DeleteOptionFunc) error {
+	return c.c.Delete(ctx, obj, opts...)
+}
+
+// Update updates the given obj in the Kubernetes cluster. It also updates the
+// fields corresponding to the status subresource for the given obj. obj must be
+// a struct pointer so that obj can be updated with the content returned by the
+// Server.
+func (c *Client) Update(ctx context.Context, obj runtime.Object) error {
+	if err := c.c.Update(ctx, obj); err != nil {
+		return err
+	}
+	return c.c.Status().Update(ctx, obj)
```

Comment:
This may result in "Stale object update" error during the second update. 
I guess, in this case, chances a low, due to "back-to-back" execution, however, on a "busy" api server, the second update has a very good chance to fail.

Reply:
Can you elaborate on why?

Reply:
https://github.com/crossplaneio/crossplane/issues/395

Reply:
Here is a fresh exmple:
```
2019-05-10T17:42:23.878-0700	ERROR	crossplane.controller.cloudsqlinstances.database.gcp.crossplane.io	failed to update status of instance	{"instance": {"kind":"CloudsqlInstance","apiVersion":"database.gcp.crossplane.io/v1alpha1","metadata":{"name":"postgresql-3be431af-71d0-11e9-8875-9cb6d08bde99","namespace":"crossplane-system","selfLink":"/apis/database.gcp.crossplane.io/v1alpha1/namespaces/crossplane-system/cloudsqlinstances/postgresql-3be431af-71d0-11e9-8875-9cb6d08bde99","uid":"3be4f49e-71d0-11e9-8875-9cb6d08bde99","resourceVersion":"316401","generation":10,"creationTimestamp":"2019-05-08T20:31:18Z","ownerReferences":[{"apiVersion":"storage.crossplane.io/v1alpha1","kind":"PostgreSQLInstance","name":"demo","uid":"3be431af-71d0-11e9-8875-9cb6d08bde99"}],"finalizers":["finalizercloudsqlinstances.database.gcp.crossplane.io"]},"spec":{"region":"us-west2","storageType":"PD_SSD","storageGB":10,"databaseVersion":"POSTGRES_9_6","tier":"db-custom-1-3840","claimRef":{"kind":"PostgreSQLInstance","namespace":"crossplane-system","name":"demo","uid":"3be431af-71d0-11e9-8875-9cb6d08bde99","apiVersion":"storage.crossplane.io/v1alpha1"},"classRef":{"kind":"ResourceClass","namespace":"crossplane-system","name":"standard-gcp-postgres","uid":"2593df83-704e-11e9-8989-9cb6d08bde99","apiVersion":"core.crossplane.io/v1alpha1","resourceVersion":"59460"},"providerRef":{"name":"demo-gcp"},"connectionSecretRef":{"name":"postgresql-3be431af-71d0-11e9-8875-9cb6d08bde99"},"reclaimPolicy":"Delete"},"status":{"Conditions":[{"Type":"Creating","Status":"False","LastTransitionTime":"2019-05-08T20:31:21Z","Reason":"instance state changed","Message":"cloud sql instance postgresql-3be431af-71d0-11e9-8875-9cb6d08bde99 is in the Creating state"},{"Type":"Ready","Status":"True","LastTransitionTime":"2019-05-08T20:35:38Z","Reason":"instance state changed","Message":"cloud sql instance postgresql-3be431af-71d0-11e9-8875-9cb6d08bde99 is in the Ready state"}],"state":"RUNNABLE","message":"Cloud SQL instance postgresql-3be431af-71d0-11e9-8875-9cb6d08bde99 is running","providerID":"https://www.googleapis.com/sql/v1beta4/projects/long-disk-191221/instances/cloudsql-3be4f49e-71d0-11e9-8875-9cb6d08bde99","endpoint":"35.236.87.69","instanceName":"cloudsql-3be4f49e-71d0-11e9-8875-9cb6d08bde99"}}, "error": "failed to update status of CRD instance postgresql-3be431af-71d0-11e9-8875-9cb6d08bde99: Operation cannot be fulfilled on cloudsqlinstances.database.gcp.crossplane.io \"postgresql-3be431af-71d0-11e9-8875-9cb6d08bde99\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/crossplaneio/crossplane/vendor/github.com/go-logr/zapr.(*zapLogger).Error
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/github.com/go-logr/zapr/zapr.go:128
github.com/crossplaneio/crossplane/pkg/controller/gcp/database.(*Reconciler).Reconcile
	/home/illya/go/src/github.com/crossplaneio/crossplane/pkg/controller/gcp/database/cloudsql_instance.go:214
github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait.Until
	/home/illya/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
```

Reply:
The issue you linked suggests the issue is caused by parallel updates; my guess being that a race causes one client to try update an object whose resource version is stale because the other client won the race and updated it first?

In this case we update in serial. I was under the impression that the Update method refreshed the object, and thus that executing two updates serially should work because the first Update refreshes the object with the latest resource version for the second Update to use.

That said, I could see a case that more frequent updates means more frequent races on a busy cluster. Is that what you're getting at here?

Reply:
I believed that myself too @negz and I'd love if we had confirmation of it.  When you call `Update`, does the `ObjectMeta.ResourceVersion` field on the object get updated? Should be easy to check.

Reply:
I was able to verify that `ObjectMeta.ResourceVersion` is updated after every `Update` or `Status.Update()`, which indeed an indication of a race condition. However, I was not able to reproduce the race condition in the code. The concurrency issue is also very weird since we don't use any concurrent processing (explicitly) in our controllers.

Reply:
Thanks for testing this @ichekrygin! I'm going to resolve this comment thread; let's move the discussion back to crossplaneio/crossplane#395.

I've removed this `Combined` client for the time being, since the Helm logic is no longer a separate controller and I wanted to integrate with your established patterns in the Gitlab controller.